ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseUri.java

```diff
@@ -76,48 +65,58 @@ private ParseUri() {
     return map;
   }
 
-  @Override
-  protected PubsubMessage processElement(PubsubMessage message)
-      throws InvalidUriException, IOException {
-    message = PubsubConstraints.ensureNonNull(message);
-    // Copy attributes
-    final Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
-    byte[] payload = message.getPayload();
-
-    // parse uri based on prefix
-    final String uri = attributes.get("uri");
-    if (uri == null) {
-      // We should only have a missing uri attribute if we're replaying messages from decoded
-      // payloads in which case they already have parsed URI attributes encoded in the payload
-      // and these will be recovered in ParsePayload.
-      return message;
-    } else if (uri.startsWith(TELEMETRY_URI_PREFIX)) {
-      // We don't yet have access to the version field, so we delay populating the document_version
-      // attribute until the ParsePayload step where we have map-like access to the JSON content.
-      attributes.put(Attribute.DOCUMENT_NAMESPACE, TELEMETRY);
-      attributes.putAll(zip(TELEMETRY_URI_SUFFIX_ELEMENTS,
-          uri.substring(TELEMETRY_URI_PREFIX.length()).split("/")));
-    } else if (uri.startsWith(GENERIC_URI_PREFIX)) {
-      attributes.putAll(
-          zip(GENERIC_URI_SUFFIX_ELEMENTS, uri.substring(GENERIC_URI_PREFIX.length()).split("/")));
-    } else if (uri.startsWith(StubUri.PREFIX)) {
-      payload = StubUri.parse(uri, attributes);
-
-      if (attributes.get(Attribute.MESSAGE_ID) != null) {
-        // convert PubSub message ID to document ID which will be a V3 UUID using a null namespace
-        // See https://stackoverflow.com/a/55296637
-        UUID documentId = UUID.nameUUIDFromBytes(
-            attributes.get(Attribute.MESSAGE_ID).getBytes(StandardCharsets.UTF_8));
-        attributes.put(Attribute.DOCUMENT_ID, documentId.toString().toLowerCase());
+  /** Factory method to create mapper instance. */
+  public static MapWithFailures<PubsubMessage, PubsubMessage, PubsubMessage> of() {
+    return MapElements.into(TypeDescriptor.of(PubsubMessage.class)).via((PubsubMessage message) -> {
+      message = PubsubConstraints.ensureNonNull(message);
+      // Copy attributes
+      final Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
+      byte[] payload = message.getPayload();
+
+      // parse uri based on prefix
+      final String uri = attributes.get("uri");
```

Comment:
```suggestion
      final String uri = attributes.get(Attributes.URI);
```

If you're open to some additional cleanup as part of this PR.